lib/finch/http1/conn.ex

```diff
@@ -100,6 +100,90 @@ defmodule Finch.HTTP1.Conn do
     end
   end
 
+  def stream(%{mint: nil} = conn, _, _, _, _, _), do: {:error, conn, "Could not connect"}
+
+  def stream(conn, req, name, ref, handler, receive_timeout, request_timeout, idle_time) do
+    full_path = Finch.Request.request_path(req)
+
+    metadata = %{request: req, name: name}
+
+    extra_measurements = %{idle_time: idle_time}
+
+    start_time = Telemetry.start(:send, metadata, extra_measurements)
+
+    try do
+      case Mint.HTTP.request(
+             conn.mint,
+             req.method,
+             full_path,
+             req.headers,
+             stream_or_body(req.body)
+           ) do
+        {:ok, mint, mint_ref} ->
+          case maybe_stream_request_body(mint, mint_ref, req.body) do
+            {:ok, mint} ->
+              Telemetry.stop(:send, start_time, metadata, extra_measurements)
+              start_time = Telemetry.start(:recv, metadata, extra_measurements)
+              resp_metadata = %{status: nil, headers: [], trailers: []}
+              timeouts = %{receive_timeout: receive_timeout, request_timeout: request_timeout}
+
+              stream =
+                Stream.resource(
+                  fn -> {[], mint, mint_ref, timeouts, resp_metadata} end,
+                  &receive_stream_response/1,
+                  fn
+                    {mint, error, resp_metadata} ->
+                      conn = %{conn | mint: mint}
+
+                      state =
+                        if open?(conn) do
+                          {:ok, conn}
+                        else
+                          :closed
+                        end
+
+                      if error do
+                        metadata = Map.merge(metadata, Map.put(resp_metadata, :error, error))
+                        Telemetry.stop(:recv, start_time, metadata, extra_measurements)
+                        raise error
+                      else
+                        metadata = Map.merge(metadata, resp_metadata)
+                        Telemetry.stop(:recv, start_time, metadata, extra_measurements)
+                      end
+
+                      send(handler, {ref, :stop, state})
```

Comment:
I don't see the connection being transferred back to the pool. Isn't that potentially an issue?

Reply:
Also notice that the after callback can be invoked in different situations:

1. The stream failed
2. Someone halted the stream

You need to recognize both scenarios accordingly. In the second one, the server can still be writing to the socket, and you need to cancel and stop that accordingly.

Reply:
Nice catch, will fix

Reply:
> In the second one, the server can still be writing to the socket, and you need to cancel and stop that accordingly.

Yes, connection is closed right below the comment `# In case some exception occured, we close the connection`

Reply:
Halting the stream also hits this branch, afaik

Reply:
The only case when connection is not closed, is them `{:done, mint_ref}` is returned by Mint. But assume that after this message, server is never writing to the socket